cpp/include/ray/api/ray_register.h

```diff
@@ -0,0 +1,201 @@
+};
+
+class Router {
+ public:
+  static Router &Instance() {
+    static Router instance;
+    return instance;
+  }
+
+  template <typename Function>
+  bool RegisterHandler(std::string const &name, const Function &f) {
+    auto pair = func_ptr_to_key_map_.emplace(AddressOf<uintptr_t>(f), name);
+    if (!pair.second) {
+      return false;
+    }
+
+    RegisterNonmemberFunc(name, f);
+    return true;
+  }
+
+  std::string Route(const char *data, std::size_t size) {
+    std::string result;
+    do {
+      try {
+        auto p = Serializer::Deserialize<std::tuple<std::string>>(data, size);
+        auto &func_name = std::get<0>(p);
+        auto it = map_invokers_.find(func_name);
+        if (it == map_invokers_.end()) {
+          result = PackArgsStr(ErrorCode::FAIL, "unknown function: " + func_name);
+          break;
+        }
+
+        it->second(data, size, result);
+        if (result.size() >= MAX_BUF_LEN) {
+          result = PackArgsStr(
+              ErrorCode::FAIL,
+              "the response result is out of range: more than 10M " + func_name);
+        }
+      } catch (const std::exception &ex) {
+        result = PackArgsStr(ErrorCode::FAIL, ex.what());
+      }
+    } while (0);
+
+    return result;
+  }
+
+ private:
+  Router() = default;
+  ~Router() = default;
+  Router(const Router &) = delete;
+  Router(Router &&) = delete;
+
+  template <typename F, size_t... I, typename Arg, typename... Args>
+  static typename std::result_of<F(Args...)>::type CallHelper(
+      const F &f, const absl::index_sequence<I...> &, std::tuple<Arg, Args...> tup) {
+    return f(std::move(std::get<I + 1>(tup))...);
+  }
+
+  template <typename F, typename Arg, typename... Args>
+  static typename std::enable_if<
+      std::is_void<typename std::result_of<F(Args...)>::type>::value>::type
+  Call(const F &f, std::string &result, std::tuple<Arg, Args...> tp) {
+    CallHelper(f, absl::make_index_sequence<sizeof...(Args)>{}, std::move(tp));
+    result = PackArgsStr(ErrorCode::OK);
+  }
+
+  template <typename F, typename Arg, typename... Args>
+  static typename std::enable_if<
+      !std::is_void<typename std::result_of<F(Args...)>::type>::value>::type
+  Call(const F &f, std::string &result, std::tuple<Arg, Args...> tp) {
+    auto r = CallHelper(f, absl::make_index_sequence<sizeof...(Args)>{}, std::move(tp));
+    result = PackArgsStr(ErrorCode::OK, "ok", r);
+  }
+
+  template <typename Function>
+  struct invoker {
+    static inline void Apply(const Function &func, const char *data, size_t size,
+                             std::string &result) {
+      using args_tuple = AddType_t<std::string, boost::callable_traits::args_t<Function>>;
+
+      try {
+        auto tp = Serializer::Deserialize<args_tuple>(data, size);
+        Router::Call(func, result, std::move(tp));
+      } catch (msgpack::type_error &e) {
+        result =
+            PackArgsStr(ErrorCode::FAIL, std::string("invalid arguments: ") + e.what());
```

Comment:
Why we don't use C++ exceptions?

Reply:
msgpack::type_error is inherited from std::exception, the exactly exception type make the error more clear.

Reply:
I mean it returns an object called `result` here. Why not throw an exception?

Reply:
I prefer doing encapsulation within the function, make it more independent, the user can get the result and return it back to the driver directly, don't need handle exception again.

Reply:
We've discussed this when @SongGuyang was about to implement the C++ worker. We agreed to use C++ exceptions instead of error codes for C++ worker API errors. @SongGuyang Does the current implementation follow this rule?

Reply:
I have did exception internal, this will return a final result and transfer back to the driver, the driver will get the result and parse it, the driver can throw ray exception when remote call failed, there is no conflict with current implementation.

Reply:
@kfstorm Yes, we prefer to use exceptions. @qicosmos We should throw a `RayException` to users directly, not a error code.

Reply:
Yes, i know that, this code return a buffer to the driver, the driver will check the result, if failed will throw ray exception, so there is no conflict here.

Reply:
Gotcha! So this function is not exposed to users. So another suggestion: replace `error_code` and `error_msg` with `ray::Status`.

Reply:
I looked at the ray::Status it lacks of some status of invoker, such as arguments not match, unknown function etc, further more this status is used internal,  the driver can still translate the deserialized result to ray::Status, so here i think don't need use ray::Status here.